# Sentiment Analysis Using TF-IDF (Term Frequency-Inverse Document Frequency) and Decision Tree

## Project Overview

Natural Language Processing (NLP) enables machines to understand, analyze, and extract meaningful information from human language. One of the most fundamental NLP tasks is **Sentiment Analysis**, which aims to determine whether a piece of text expresses a positive or negative opinion.

In this project, I build a complete sentiment analysis pipeline using classical NLP techniques and Machine Learning algorithms. The workflow covers the entire process from raw text preprocessing to feature extraction, model training, evaluation, and experiment documentation.

The primary objective of this project is not only to achieve high predictive performance but also to systematically investigate how different preprocessing techniques, feature engineering strategies, and machine learning models affect sentiment classification performance.

---

## Dataset

Dataset Link: **[IMBD Movie Reviews](https://www.kaggle.com/datasets/mwallerphunware/imbd-movie-reviews-for-binary-sentiment-analysis)**

The dataset consists of movie reviews labeled with their corresponding sentiment:

* Positive
* Negative

The reviews contain real-world natural language, making them suitable for evaluating text preprocessing techniques and machine learning models.

---

## Project Pipeline

The following stages are implemented throughout this project:

### 1. Text Preprocessing

* Contraction expansion
* Lowercasing
* Text cleaning using regular expressions
* Stopword removal with preserved negations
* Lemmatization
* Corpus construction

### 2. Feature Extraction

* Term Frequency-Inverse Document Frequency (TfidfVectorizer)
* N-gram generation
* Vocabulary analysis
* Feature selection using `max_features`
* Rare-word filtering using `min_df`

### 3. Model Training

Different machine learning algorithms are evaluated and compared, including:

* Gaussian Naive Bayes
* Multinomial Naive Bayes
* Bernoulli Naive Bayes
* Logistic Regression
* Support Vector Machines (SVM)
* Decision Trees
* Random Forests

### 4. Model Evaluation

Performance is assessed using multiple metrics:

* Accuracy
* Precision
* Recall
* F1-Score
* ROC-AUC Score
* Confusion Matrix
* Cross-Validation Mean Accuracy
* Cross-Validation Standard Deviation

---

## Experimental Approach

Rather than training a single model, this notebook follows an experimentation-driven methodology.

For each model, multiple configurations are tested, including different values for:

* `max_features`
* `min_df`
* N-gram ranges
* Preprocessing strategies

The goal is to identify the most effective configuration while understanding the trade-offs between model complexity, computational cost, and predictive performance.

---

## Key Learning Objectives

Through this project, I aim to:

* Develop a deeper understanding of NLP preprocessing techniques.
* Compare the behavior of different machine learning algorithms on text data.
* Analyze how feature engineering impacts classification performance.
* Build reproducible NLP pipelines suitable for real-world applications.
* Establish strong baselines before moving toward advanced embedding and transformer-based approaches.

---

**Author:** Hazem Mohamed

**Role:** AI Engineer | Machine Learning Engineer | NLP Engineer

**Repository:** [NLP Experimentation Lab](https://github.com/Hazem1695/NLP-Experimentation-Lab)


# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

## **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv('MovieReviewTrainingDatabase.csv')
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  25000 non-null  object
 1   review     25000 non-null  object
dtypes: object(2)
memory usage: 390.8+ KB
None

SHAPE OF DATA
Rows: 25000, Columns: 2

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 96

DATA TYPES
sentiment    object
review       object
dtype: object

NUMERICAL SUMMARY
       sentiment                                             review
count      25000                                              25000
unique         2                                              24904
top     Positive  You do realize that you've been watching the E...
freq       12500                                                  3

CATEGORICAL SUMMARY
       sentiment                                             review
count      25000                 

## Duplicate Data Detection

In [4]:
duplicates = dataset[dataset.duplicated(subset=['review'], keep=False)]
duplicates.sort_values('review')

,sentiment,review
21186,Negative,"Back in his youth, the old man had wanted to..."
21877,Negative,"Back in his youth, the old man had wanted to..."
14734,Negative,'Dead Letter Office' is a low-budget film abou...
5519,Negative,'Dead Letter Office' is a low-budget film abou...
7011,Positive,".......Playing Kaddiddlehopper, Col San Fernan..."
...,...,...
2685,Negative,"in this movie, joe pesci slams dunks a basketb..."
22244,Positive,it's amazing that so many people that i know h...
14767,Positive,it's amazing that so many people that i know h...
12462,Negative,this movie begins with an ordinary funeral... ...


## Quantifying Duplicate Review Frequencies

In [5]:
review_counts = dataset['review'].value_counts()
print("Reviews appearing more than once:")
print((review_counts > 1).sum())
print("\nMaximum repetitions:")
print(review_counts.max())

Reviews appearing more than once:
92

Maximum repetitions:
3


## Removing Duplicate Reviews & Resetting Index
> **Note:** This cell drops the repeated rows we identified in the previous steps and cleanly resets the row indices for model training

In [6]:
print("Before:", len(dataset))
dataset = dataset.drop_duplicates()
print("After:", len(dataset))
dataset = dataset.reset_index(drop=True)

Before: 25000
After: 24904


## Library Installation
> **Note:** The `contractions` library is required to automatically expand shortcuts like *don't* to *do not* and *I'm* to *I am* during preprocessing.

In [7]:
!pip install contractions

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 7.3 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 6.4 MB/s eta 0:00:00


## **Cleaning the texts**

In [8]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')
import re
import contractions
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

base_stopwords = set(stopwords.words('english'))
negation_words = {'not', 'no', 'never'}
all_stopwords = base_stopwords - negation_words

wnl = WordNetLemmatizer()

corpus = []

for i in range(0, len(dataset)):
    review = dataset['review'][i]
    # Fix contractions
    review = contractions.fix(review)
    # Lowercase
    review = review.lower()
    
    review = re.sub(r'[^a-zA-Z\s]', ' ', review)
    # Split
    words = review.split()
    # Chained Lemmatization (Handles both Verbs 'v' and Nouns 'n')
    review = [wnl.lemmatize(wnl.lemmatize(word, pos='v'), pos='n') for word in words if word not in all_stopwords]
    review = ' '.join(review) 
    corpus.append(review)

[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /usr/share/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /usr/share/nltk_data...


## Preprocessing Verification
> **Note:** Pulling the first two rows directly as a memory array to confirm that our lowercasing, stopword stripping, and lemmatization pipeline worked correctly before feeding it into the vectorizer.

In [9]:
# Pull the data directly as a fast memory array
raw_samples = dataset['review'].head(2).values

for i in range(2):
    print(f"=== REVIEW #{i+1} ===")
    print(f"RAW:     {raw_samples[i]}\n") 
    print(f"CLEANED: {corpus[i]}")
    print("-" * 50)

=== REVIEW #1 ===
RAW:     With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.  Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.  The actual feature film bit when it final

# **Encoding Categorical data Using Label Encoding**

In [10]:
from sklearn.preprocessing import LabelEncoder
y = dataset.iloc[:, 0].values
le = LabelEncoder()
y = le.fit_transform(y)

In [11]:
print(y)

[1 1 0 ... 0 0 1]


# Class Balance Check
> **Note:** Using NumPy to verify if our dataset is perfectly balanced between positive and negative reviews before splitting it into training and testing sets.

In [12]:
# This returns the unique classes and how many times they appear
classes, counts = np.unique(y, return_counts=True)
for c, count in zip(classes, counts):
    print(f"Class {c} contains {count}")

Class 0 contains 12432
Class 1 contains 12472


# **Splitting the dataset into the Training set and Test set**

In [13]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(corpus, y, test_size = 0.20, random_state = 0)

# **Creating the TF-IDF (Term Frequency-Inverse Document Frequency) model**

In [14]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Add ngram_range=(1, 2) so it automatically catches phrases like "not good" or "no clue"
tf = TfidfVectorizer(max_features=10000, ngram_range=(1,2))
X_train = tf.fit_transform(X_train)
X_test = tf.transform(X_test)

In [15]:
print(X_train.shape)        # (n_samples, n_features)
print(X_train.shape[0])     # number of training samples
print(X_train.shape[1])     # number of features (vocabulary size)

(19923, 10000)
19923
10000


In [16]:
print(X_test.shape)        # (n_samples, n_features)
print(X_test.shape[0])     # number of training samples
print(X_test.shape[1])     # number of features (vocabulary size)

(4981, 10000)
4981
10000


In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(random_state=0)

param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'splitter': ['best', 'random'],
    'max_depth': [None, 5, 10, 20, 30, 50],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 4, 8],
    'max_features': [None, 'sqrt', 'log2']
}

In [ ]:
random_search = RandomizedSearchCV(
    estimator=dt,
    param_distributions=param_grid,
    n_iter=30,          # Number of random combinations
    scoring='accuracy',
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2
)

In [19]:
random_search.fit(X_train, y_train)

print(random_search.best_params_)
print(random_search.best_score_)

Fitting 5 folds for each of 30 candidates, totalling 150 fits
[CV] END criterion=entropy, max_depth=None, max_features=None, min_samples_leaf=4, min_samples_split=20, splitter=random; total time=  10.9s
[CV] END criterion=log_loss, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, splitter=random; total time=   0.3s
[CV] END criterion=entropy, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=10, splitter=best; total time=   0.4s
[CV] END criterion=entropy, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=10, splitter=best; total time=   0.3s
[CV] END criterion=log_loss, max_depth=20, max_features=sqrt, min_samples_leaf=8, min_samples_split=5, splitter=best; total time=   0.1s
[CV] END criterion=log_loss, max_depth=20, max_features=sqrt, min_samples_leaf=8, min_samples_split=5, splitter=best; total time=   0.2s
[CV] END criterion=log_loss, max_depth=5, max_features=None, min_samples_leaf=2, min_samples_split=20, s

# **Training the Decision Tree model on the Training set**

In [17]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(splitter='random', max_depth=20, min_samples_split=2, min_samples_leaf=2, criterion='gini', random_state=0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=20, min_samples_leaf=2, random_state=0,
                       splitter='random')

# **Predicting the Test set results**

In [18]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [0 1]
 [1 0]
 ...
 [0 0]
 [0 0]
 [1 0]]


# **Evaluating the Model Performance**

In [19]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=3)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())

Confusion Matrix:
[[1698  822]
 [ 426 2035]]

Accuracy Score:
0.7494479020277053

Classification Report:
              precision    recall  f1-score   support

           0       0.80      0.67      0.73      2520
           1       0.71      0.83      0.77      2461

    accuracy                           0.75      4981
   macro avg       0.76      0.75      0.75      4981
weighted avg       0.76      0.75      0.75      4981


ROC-AUC Score:
0.750354579052263

Mean Accuracy:
0.7287055162375143

Standard Deviation:
0.0024259490987491136


# Decision Tree Performance Analysis for Text Classification (TF-IDF)

# 1. Objective

The objective of this experiment was to evaluate the effectiveness of the **Decision Tree** algorithm for binary text classification using features generated by **TfidfVectorizer**.

Unlike linear models and probabilistic classifiers, Decision Trees partition the feature space through a sequence of hierarchical decisions. Although Decision Trees are highly interpretable, they are often challenged by high-dimensional and sparse text representations. This study investigates how feature extraction, tree complexity, and hyperparameter optimization influence their performance under TF-IDF weighting.

The experiment aims to answer the following research questions:

* How well does Decision Tree perform on sparse, TF-IDF-weighted text data?
* Does increasing the vocabulary size improve performance?
* Can hyperparameter optimization significantly improve the classifier?
* Why was **RandomizedSearchCV** selected instead of **GridSearchCV**?
* How does the Decision Tree compare against previously evaluated TF-IDF models?

---

# 2. Experimental Setup

## Dataset

* Final test set: **4,981 documents** (2,520 negative / 2,461 positive).

## Feature Extraction

Documents were transformed into numerical vectors using **TfidfVectorizer**.

| Configuration | Parameters                            |
| -------------- | -------------------------------------- |
| Baseline       | `TfidfVectorizer()` (default settings) |
| C1             | max_features=5000, ngram_range=(1,2)   |
| C2             | max_features=10000, ngram_range=(1,2)  |
| C3             | max_features=15000, ngram_range=(1,2)  |

---

# 3. Hyperparameter Optimization Strategy

**GridSearchCV** was considered first for hyperparameter optimization.

The complete search space contained:

* **1,728 parameter combinations**
* **5-fold Cross Validation**

Resulting in:

```text
1,728 × 5 = 8,640 model trainings
```

Training thousands of Decision Trees on high-dimensional text data would require a considerable amount of computation time.

Instead, **RandomizedSearchCV** was selected.

Configuration:

```text
30 random parameter combinations
5-fold Cross Validation
```

Resulting in only:

```text
30 × 5 = 150 model trainings
```

This reduced the computational cost by approximately **98%** (150 vs. 8,640 fits) while exploring only **~1.7%** of the full grid — RandomizedSearchCV was chosen because it is substantially faster than GridSearchCV while still surfacing strong configurations.

---

# 4. Hyperparameters Explored

The optimization process searched over parameters including:

* Tree depth (`max_depth`)
* Splitting strategy (`splitter`)
* Splitting criterion (`gini`, `entropy`)
* Minimum samples required for splitting (`min_samples_split`)
* Minimum samples required per leaf (`min_samples_leaf`)
* Feature selection strategy (`max_features`)

The best configurations discovered were:

### 5,000 Features

```python
DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=30,
    min_samples_split=10,
    min_samples_leaf=8
)
```

### 10,000 Features

```python
DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2
)
```

### 15,000 Features

The same configuration as 10,000 features remained optimal, indicating the search converged on a stable structure once the vocabulary was large enough.

---

# 5. Experimental Results

| Configuration              | Accuracy   | ROC-AUC    | CV Mean | CV Std |
| --------------------------- | ---------- | ---------- | ------- | ------ |
| Baseline (default, no tune) | 70.53%     | 0.7053     | 70.36%  | 0.0030 |
| RandomizedSearch (5K)       | 71.99%     | 0.7201     | 71.32%  | 0.0025 |
| RandomizedSearch (10K)      | **74.94%** | **0.7504** | 72.87%  | 0.0024 |
| RandomizedSearch (15K)      | 73.98%     | 0.7404     | 73.10%  | 0.0065 |

---

# 6. Performance Analysis

## Effect of Hyperparameter Optimization

Hyperparameter tuning produced a clear improvement over the untuned baseline.

Accuracy increased from:

**70.53%** (baseline)

to

**74.94%** (best tuned model, 10K features)

representing an improvement of over **4 percentage points**.

## Why RandomizedSearchCV?

| Method             | Candidates | Total Fits |
| ------------------- | ---------: | ---------: |
| GridSearchCV        |      1,728 |      8,640 |
| RandomizedSearchCV  |         30 |        150 |

RandomizedSearchCV evaluated only **1.7%** of the total parameter combinations yet discovered a configuration that meaningfully outperformed the untuned baseline, at roughly **2% of the computational cost** of an exhaustive grid search.

## Effect of Vocabulary Size

| Features | Accuracy   |
| -------- | ---------- |
| 5K       | 71.99%     |
| **10K**  | **74.94%** |
| 15K      | 73.98%     |

Accuracy increased sharply from 5K to 10K features, then declined slightly at 15K. This is a cleaner "rise then plateau/decline" pattern than what was observed in the earlier CountVectorizer-based Decision Tree experiment (where 5K unexpectedly edged out 10K). Under TF-IDF weighting, 10K appears to be the sweet spot — enough vocabulary to capture useful splits, without diluting the tree with too many low-signal, high-dimensional features.

## Effect of Splitting Strategy

The optimized models consistently selected:

```python
splitter='random'
```

rather than `splitter='best'`. Random splitting introduces additional randomness during tree construction, which reduces overfitting and tends to generalize better on high-dimensional sparse data, where greedily choosing the locally optimal split can produce overly specific decision boundaries.

---

# 7. Precision and Recall Analysis

### Baseline (untuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.71      | 0.71   | 0.71     |
| 1     | 0.70      | 0.70   | 0.70     |

### 5K Features (Tuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.73      | 0.71   | 0.72     |
| 1     | 0.71      | 0.73   | 0.72     |

### Best Model — 10K Features (Tuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.80      | 0.67   | 0.73     |
| 1     | 0.71      | 0.83   | 0.77     |

### 15K Features (Tuned)

| Class | Precision | Recall | F1-score |
| ----- | --------- | ------ | -------- |
| 0     | 0.77      | 0.69   | 0.73     |
| 1     | 0.71      | 0.79   | 0.75     |

The best model (10K features) shows the sharpest precision/recall asymmetry of the group: very high precision on Class 0 (0.80) paired with comparatively low recall (0.67), against high recall on Class 1 (0.83). In practice this means the tree is quite cautious about calling something Class 0 — when it does, it's usually right — but it misses a third of the true Class 0 cases, folding many of them into Class 1 instead.

---

# 8. ROC-AUC Analysis

| Configuration           | ROC-AUC    |
| ------------------------ | ---------- |
| Baseline (untuned)       | 0.7053     |
| RandomizedSearch (5K)    | 0.7201     |
| **RandomizedSearch (10K)**| **0.7504** |
| RandomizedSearch (15K)   | 0.7404     |

ROC-AUC follows the same rise-then-decline pattern as accuracy, confirming 10K features as the strongest class-separation point among the configurations tested.

---

# 9. Cross-Validation Analysis

| Configuration           | CV Mean Accuracy | CV Std Dev |
| ------------------------ | ----------------- | ---------- |
| Baseline (untuned)       | 70.36%             | 0.0030     |
| RandomizedSearch (5K)    | 71.32%             | 0.0025     |
| **RandomizedSearch (10K)**| **72.87%**         | **0.0024** |
| RandomizedSearch (15K)   | 73.10%             | 0.0065     |

The 10K configuration offers the best balance of accuracy and stability — it has both a strong mean and the tightest standard deviation of the tuned runs. The 15K configuration has a marginally higher CV mean but a notably larger spread (0.0065 vs. 0.0024), meaning its performance was less consistent across folds despite scoring similarly on the held-out test set.

---

# 10. Best Model Configuration

The highest-performing Decision Tree was:

```python
TfidfVectorizer(
    max_features=10000,
    ngram_range=(1, 2)
)

DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2,
    random_state=0
)
```

Performance:

* Accuracy = **74.94%**
* Precision (Class 0 / 1) = **0.80 / 0.71**
* Recall (Class 0 / 1) = **0.67 / 0.83**
* ROC-AUC = **0.7504**
* Cross-Validation Accuracy = **72.87%**
* Cross-Validation Standard Deviation = **0.0024**

---

# 11. Comparison with Previously Evaluated Models (TF-IDF)

| Model                    | Best Accuracy | ROC-AUC    |
| -------------------------- | -------------- | ---------- |
| **LinearSVC**               | **90.18%**     | **0.9020** |
| Logistic Regression         | 89.46%         | 0.8948     |
| Naive Bayes (Multinomial)   | 88.13%         | 0.8815     |
| K-Nearest Neighbors         | 80.41%         | 0.8047     |
| **Decision Tree**           | **74.94%**     | **0.7504** |

Decision Tree remains the weakest of the evaluated models under TF-IDF weighting, trailing K-Nearest Neighbors by roughly 5.5 points and LinearSVC by over 15 points. Its ranking relative to the other classifiers is unchanged from the CountVectorizer experiment, though its absolute accuracy did improve under TF-IDF (74.94% vs. 74.34% best BoW result).

---

# 12. Discussion

The experiments confirm the general limitation of a single Decision Tree for sparse, high-dimensional text classification, even under TF-IDF weighting: axis-aligned splits still struggle to build compact, generalizable boundaries over thousands of largely-zero dimensions.

RandomizedSearchCV again proved to be a practical substitute for GridSearchCV, cutting required trainings from 8,640 to 150 (~98% reduction) while still finding configurations that meaningfully improved accuracy over the untuned baseline.

Unlike the earlier CountVectorizer-based run — where 5K features unexpectedly outperformed 10K — this TF-IDF run showed the more intuitive pattern of accuracy rising to a peak at 10K features before declining slightly at 15K. This suggests TF-IDF's down-weighting of common terms makes the tree's splits more informative as vocabulary grows, at least up to a point, whereas raw counts (BoW) let the tree's outcome depend more on incidental noise in which of the larger feature set happened to get selected.

Across both feature extraction methods, Decision Tree remains substantially behind linear and probabilistic models, reinforcing that a single tree is not well suited to high-dimensional sparse text representations regardless of whether counts or TF-IDF weighting are used.

---

# 13. Final Conclusion

This experiment evaluated Decision Tree performance on binary text classification using TF-IDF features across four configurations, with RandomizedSearchCV used for hyperparameter tuning.

The best-performing model used **TfidfVectorizer with 10,000 features and bigrams**, paired with an optimized **DecisionTreeClassifier** (random splitting, Gini impurity, max_depth=20), achieving **74.94% accuracy** and **0.7504 ROC-AUC** — its strongest result across either feature extraction method tried so far.

Compared against previously evaluated TF-IDF classifiers, Decision Tree ranks last, well behind LinearSVC (90.18%), Logistic Regression (89.46%), Multinomial Naive Bayes (88.13%), and KNN (80.41%). These results reinforce that single Decision Trees, while interpretable, are not the most suitable choice for high-dimensional sparse text classification — ensemble methods such as Random Forest are the natural next step to test whether combining multiple trees closes this gap.